File: DS-lab/worksheet-6/prob1.java
import java.util.*;

class minStack {
    Stack<Integer> s1;
    Stack<Integer> s2;
    
    public minStack() {
        s1 = new Stack<>();
        s2 = new Stack<>();
    }
    
    public void push(int x) {
        s1.push(x);
        if (s2.isEmpty() || s2.peek() >= x) {
            s2.push(x);
        }
    }
    
    public int pop() {
        if (s1.isEmpty()) {
            System.out.println("Stack is empty");
            return -1;
        }
        int val = s1.pop();
        if (!s2.isEmpty() && val == s2.peek()) {
            s2.pop();
        }
        System.out.println("Popped: " + val);
        return val;
    }
    
    public int getMin() {
        if (s2.isEmpty()) {
            System.out.println("Stack is empty, no minimum");
            return -1;
        }
        System.out.println("Min: " + s2.peek());
        return s2.peek();
    }
}

public class Main {
    public static void main(String[] args) {
        minStack s = new minStack();
        s.push(3);
        s.push(2);
        s.push(10);
        s.push(1);
        s.getMin();  // Min: 1
        s.pop();     // Popped: 1
        s.getMin();  // Min: 2
        s.pop();     // Popped: 10
        s.getMin();  // Min: 2
        s.pop();     // Popped: 2
        s.getMin();  // Min: 3
        s.pop();     // Popped: 3
        s.getMin();  // Stack is empty, no minimum
    }
}

File: DS-lab/worksheet-6(practice)/prob1.c
#include <stdio.h>
#define MAX 100

typedef struct{
    int arr[MAX];
    int top;
} Stack;
void init(Stack *s){
    s->top=-1;
}
int isEmpty(Stack *s){
    return s->top==-1;
}

int isFull(Stack *s){
    return s->top==MAX-1;
}

int push(Stack *s,int ele){
    if(isFull(s)){
        printf("Stack is full\n");
    }else{
    printf("The element %d is pushed\n",ele);
    s->arr[++s->top] = ele;
    }
}

int pop(Stack *s){
    if(isEmpty(s)){
        printf("Stack is Empty\n");
        return 0;}
    printf("The element %d is popped\n",s->arr[s->top]);
    return s->arr[s->top--];
}

int peek(Stack *s){
    if(isEmpty(s)){
        printf("Stack is Empty");
        return 0;}
    printf("The element %d is at top\n",s->arr[s->top]);
    return s->arr[s->top];
}

int isContinue(){
    char isCont;
    while(getchar()!='\n');
    printf("Do you want to continue: ");
    scanf("%c",&isCont);
    if(isCont=='y'){
        return 1;
    }
    return 0;
}

int showMenu(){
    int menu;
    printf("Menu functions : \n\t 1.push\n\t 2.pop \n\t 3.peek\n");
    printf("Enter menu: ");
    scanf("%d",&menu);
    return menu;
}


int main(){
    Stack s;
    init(&s);
    int menu;
    do{
    menu = showMenu();
    if(menu == 1){
    int ele;
    printf("Enter the element\n");
    scanf("%d",&ele);
    push(&s,ele);
    }
    else if ( menu == 2){
    pop(&s);
    }
    else{
    peek(&s);
    }}
    while(isContinue());

}
